nomos-mix/core/src/message_blend/temporal.rs

```diff
@@ -1,8 +1,101 @@
+use std::{
+    collections::VecDeque,
+    pin::Pin,
+    task::{Context, Poll},
+    time::Duration,
+};
+
+use futures::{Future, Stream};
+use rand::Rng;
 use serde::{Deserialize, Serialize};
+use tokio::time;
+
+/// [`TemporalProcessor`] delays messages randomly to hide timing correlation
+/// between incoming and outgoing messages from a node.
+///
+/// See the [`Stream`] implementation below for more details on how it works.
+pub struct TemporalProcessor<M> {
+    settings: TemporalProcessorSettings,
+    // All scheduled messages
+    queue: VecDeque<M>,
+    /// Interval in seconds for running the lottery to release a message
+    lottery_interval: time::Interval,
+    /// To wait a few seconds after running the lottery before releasing the message.
+    /// The lottery returns how long to wait before releasing the message.
+    release_timer: Option<Pin<Box<time::Sleep>>>,
+}
 
 #[derive(Clone, Debug, Serialize, Deserialize)]
 pub struct TemporalProcessorSettings {
     pub max_delay_seconds: u64,
 }
 
-// TODO: Implement TemporalProcessor
+impl<M> TemporalProcessor<M> {
+    #[allow(dead_code)] // TODO: Remove this macro once this module is integrated
+    pub fn new(settings: TemporalProcessorSettings) -> Self {
+        let lottery_interval = Self::lottery_interval(settings.max_delay_seconds);
+        Self {
+            settings,
+            queue: VecDeque::new(),
+            lottery_interval,
+            release_timer: None,
+        }
+    }
+
+    /// Create [`time::Interval`] for running the lottery to release a message.
+    fn lottery_interval(max_delay_seconds: u64) -> time::Interval {
+        time::interval(Duration::from_secs(Self::lottery_interval_seconds(
+            max_delay_seconds,
+        )))
+    }
+
+    /// Calculate the interval in seconds for running the lottery.
+    /// The lottery interval is half of the maximum delay,
+    /// in order to guarantee that the interval between two subsequent message emissions
+    /// is at most [`max_delay_seconds`].
+    fn lottery_interval_seconds(max_delay_seconds: u64) -> u64 {
+        max_delay_seconds / 2
+    }
+
+    /// Run the lottery to determine the delay before releasing a message.
+    /// The delay is in [0, `lottery_interval_seconds`).
+    fn run_lottery(&self) -> u64 {
+        let interval = Self::lottery_interval_seconds(self.settings.max_delay_seconds);
+        rand::thread_rng().gen_range(0..interval)
+    }
+
+    /// Schedule a message to be released later.
+    #[allow(dead_code)] // TODO: Remove this macro once this module is integrated
+    pub fn push_message(&mut self, message: M) {
+        self.queue.push_back(message);
+    }
+}
+
+impl<M> Stream for TemporalProcessor<M>
+where
+    M: Unpin + Clone + 'static,
```

Comment:
Nice! I think this is the way. We need to think how to properly fit with the other pieces as streams as well.